Data fields
Here's a brief version of what you'll find in the data description file.

    SalePrice - the property's sale price in dollars. This is the target variable that you're trying to predict.
    MSSubClass: The building class
    MSZoning: The general zoning classification
    LotFrontage: Linear feet of street connected to property
    LotArea: Lot size in square feet
    Street: Type of road access
    Alley: Type of alley access
    LotShape: General shape of property
    LandContour: Flatness of the property
    Utilities: Type of utilities available
    LotConfig: Lot configuration
    LandSlope: Slope of property
    Neighborhood: Physical locations within Ames city limits
    Condition1: Proximity to main road or railroad
    Condition2: Proximity to main road or railroad (if a second is present)
    BldgType: Type of dwelling
    HouseStyle: Style of dwelling
    OverallQual: Overall material and finish quality
    OverallCond: Overall condition rating
    YearBuilt: Original construction date
    YearRemodAdd: Remodel date
    RoofStyle: Type of roof
    RoofMatl: Roof material
    Exterior1st: Exterior covering on house
    Exterior2nd: Exterior covering on house (if more than one material)
    MasVnrType: Masonry veneer type
    MasVnrArea: Masonry veneer area in square feet
    ExterQual: Exterior material quality
    ExterCond: Present condition of the material on the exterior
    Foundation: Type of foundation
    BsmtQual: Height of the basement
    BsmtCond: General condition of the basement
    BsmtExposure: Walkout or garden level basement walls
    BsmtFinType1: Quality of basement finished area
    BsmtFinSF1: Type 1 finished square feet
    BsmtFinType2: Quality of second finished area (if present)
    BsmtFinSF2: Type 2 finished square feet
    BsmtUnfSF: Unfinished square feet of basement area
    TotalBsmtSF: Total square feet of basement area
    Heating: Type of heating
    HeatingQC: Heating quality and condition
    CentralAir: Central air conditioning
    Electrical: Electrical system
    1stFlrSF: First Floor square feet
    2ndFlrSF: Second floor square feet
    LowQualFinSF: Low quality finished square feet (all floors)
    GrLivArea: Above grade (ground) living area square feet
    BsmtFullBath: Basement full bathrooms
    BsmtHalfBath: Basement half bathrooms
    FullBath: Full bathrooms above grade
    HalfBath: Half baths above grade
    Bedroom: Number of bedrooms above basement level
    Kitchen: Number of kitchens
    KitchenQual: Kitchen quality
    TotRmsAbvGrd: Total rooms above grade (does not include bathrooms)
    Functional: Home functionality rating
    Fireplaces: Number of fireplaces
    FireplaceQu: Fireplace quality
    GarageType: Garage location
    GarageYrBlt: Year garage was built
    GarageFinish: Interior finish of the garage
    GarageCars: Size of garage in car capacity
    GarageArea: Size of garage in square feet
    GarageQual: Garage quality
    GarageCond: Garage condition
    PavedDrive: Paved driveway
    WoodDeckSF: Wood deck area in square feet
    OpenPorchSF: Open porch area in square feet
    EnclosedPorch: Enclosed porch area in square feet
    3SsnPorch: Three season porch area in square feet
    ScreenPorch: Screen porch area in square feet
    PoolArea: Pool area in square feet
    PoolQC: Pool quality
    Fence: Fence quality
    MiscFeature: Miscellaneous feature not covered in other categories
    MiscVal: $Value of miscellaneous feature
    MoSold: Month Sold
    YrSold: Year Sold
    SaleType: Type of sale
    SaleCondition: Condition of sale

### 1.1 import some necessary librairies

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

### 1.2 import and put the train and test datasets in  pandas dataframe

In [2]:
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

### 1.3 display the first five rows of the train dataset.

In [3]:
train.head(5)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

### 1.4 display the first five rows of the test dataset.

In [4]:
test.head(5)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence MiscFeature  \
0         Lvl    AllPub  ...         120        0    NaN  MnPrv         NaN   
1         Lvl    AllPub  ...           0        0    NaN    NaN        Gar2   
2         Lvl    AllPub  ...           0        0    NaN  MnPrv         NaN   
3         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
4         HLS    AllPub  ...         144        0    NaN    NaN         NaN   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  
0       0      6    2010        WD         Normal  
1   12500      6    2010        WD         Normal  
2       0      3    2010        WD         Normal  
3       0      6    2010        WD         Normal  
4       0      1    2010        WD         Normal  

[5 rows x 80 columns]

### 1.5 check the numbers of samples and features

In [5]:
train.count()

Id               1460
MSSubClass       1460
MSZoning         1460
LotFrontage      1201
LotArea          1460
                 ... 
MoSold           1460
YrSold           1460
SaleType         1460
SaleCondition    1460
SalePrice        1460
Length: 81, dtype: int64

In [6]:
test.count()

Id               1459
MSSubClass       1459
MSZoning         1455
LotFrontage      1232
LotArea          1459
                 ... 
MiscVal          1459
MoSold           1459
YrSold           1459
SaleType         1458
SaleCondition    1459
Length: 80, dtype: int64

In [7]:
train.shape

(1460, 81)

In [8]:
test.shape

(1459, 80)

In [9]:
train.size

118260

In [10]:
test.size

116720

### 1.6 Save the 'Id' column

In [11]:
Id_train =  train['Id'].copy() 
Id_test =  test['Id'].copy()

### 1.7 Now drop the  'Id' colum since it's unnecessary for  the prediction process.

In [12]:
train = train.drop(['Id'], axis=1)
train.head(5)

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0          60       RL         65.0     8450   Pave   NaN      Reg   
1          20       RL         80.0     9600   Pave   NaN      Reg   
2          60       RL         68.0    11250   Pave   NaN      IR1   
3          70       RL         60.0     9550   Pave   NaN      IR1   
4          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig  ... PoolArea PoolQC Fence MiscFeature  \
0         Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
1         Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   
2         Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
3         Lvl    AllPub    Corner  ...        0    NaN   NaN         NaN   
4         Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
0       0      2    2008        WD         Normal     208500  
1       0      5    2007        WD         Normal     181500  
2       0      9    2008        WD         Normal     223500  
3       0      2    2006        WD        Abnorml     140000  
4       0     12    2008        WD         Normal     250000  

[5 rows x 80 columns]

In [13]:
test = test.drop(['Id'], axis=1)
test.head(5)

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0          20       RH         80.0    11622   Pave   NaN      Reg   
1          20       RL         81.0    14267   Pave   NaN      IR1   
2          60       RL         74.0    13830   Pave   NaN      IR1   
3          60       RL         78.0     9978   Pave   NaN      IR1   
4         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities LotConfig  ... ScreenPorch PoolArea PoolQC  Fence  \
0         Lvl    AllPub    Inside  ...         120        0    NaN  MnPrv   
1         Lvl    AllPub    Corner  ...           0        0    NaN    NaN   
2         Lvl    AllPub    Inside  ...           0        0    NaN  MnPrv   
3         Lvl    AllPub    Inside  ...           0        0    NaN    NaN   
4         HLS    AllPub    Inside  ...         144        0    NaN    NaN   

  MiscFeature MiscVal  MoSold  YrSold  SaleType  SaleCondition  
0         NaN       0       6    2010        WD         Normal  
1        Gar2   12500       6    2010        WD         Normal  
2         NaN       0       3    2010        WD         Normal  
3         NaN       0       6    2010        WD         Normal  
4         NaN       0       1    2010        WD         Normal  

[5 rows x 79 columns]

### 1.8 check again the data size after dropping the 'Id' variable

In [14]:
train.size

116800

In [15]:
test.size

115261

In [16]:
train.shape

(1460, 80)

In [17]:
test.shape

(1459, 79)

## 1.0 Data Processing

### concatenate the train and test data in the same dataframe

In [18]:
tt = pd.concat([train,test], sort="False")
tt.shape

(2919, 80)

### 1.1 Checking the all Missing Data?

In [19]:
dupes = tt.duplicated()
sum(dupes)

0

In [20]:
tt.isnull().values.any()

True

In [21]:
tt.isnull().sum().sum()

15424

#### observation:
   Total 15424 missing values detected in the whole dataset

In [22]:
tt_copy= tt.copy()

#### 1.15 Imputing missing values

#### PoolQC : 
data description says NA means "No Pool". That make sense, given the huge ratio of missing value (+99%) and majority of houses have no Pool at all in general.

In [23]:
tt_copy['PoolQC'].isnull().sum()

2909

In [24]:
tt_copy['PoolQC'].unique()

array([nan, 'Ex', 'Fa', 'Gd'], dtype=object)

In [25]:
tt_copy['PoolQC'] = tt_copy['PoolQC'].transform(lambda x: x.fillna("NA"))
tt_copy['PoolQC'].unique()

array(['NA', 'Ex', 'Fa', 'Gd'], dtype=object)

#### MiscFeature : 
data description says NA means "no misc feature"

In [26]:
tt_copy['MiscFeature'].isnull().sum()

2814

In [27]:
tt_copy['MiscFeature'].unique()

array([nan, 'Shed', 'Gar2', 'Othr', 'TenC'], dtype=object)

In [28]:
tt_copy['MiscFeature'] = tt_copy['MiscFeature'].transform(lambda x: x.fillna("NA"))
tt_copy['MiscFeature'].unique()

array(['NA', 'Shed', 'Gar2', 'Othr', 'TenC'], dtype=object)

#### Alley : 
data description says NA means "no alley access"

In [29]:
tt_copy['Alley'] = tt_copy['Alley'].transform(lambda x: x.fillna("NA"))
tt_copy['Alley'].unique()

array(['NA', 'Grvl', 'Pave'], dtype=object)

#### Fence : 
data description says NA means "no fence"

In [30]:
tt_copy['Fence'] = tt_copy['Fence'].transform(lambda x: x.fillna("NA"))
tt_copy['Fence'].unique()

array(['NA', 'MnPrv', 'GdWo', 'GdPrv', 'MnWw'], dtype=object)

### 1.16 Check remaining missing values if any 

In [31]:
#list(tt_copy.columns)
tt_copy.isna().sum().sum()

4632

In [32]:
# FireplaceQu : data description says NA means "no fireplace"

#tt_copy['FireplaceQu'].unique()
tt_copy['FireplaceQu'] = tt_copy['FireplaceQu'].replace(np.nan,"NA")

In [33]:
#LotFrontage : Since the area of each street connected to the house property most likely have a similar
#Group by neighborhood and fill in missing value by the median LotFrontage of all the neighborhood
tt_copy['LotFrontage'] = tt_copy.groupby('Neighborhood')['LotFrontage'].transform(lambda x: x.fillna(x.median()))
tt_copy['LotFrontage'].isna().sum()

0

In [34]:
# GarageType, GarageFinish, GarageQual and GarageCond : Replacing missing data with None

#tt_copy['GarageType'].unique()
tt_copy = tt_copy.replace({"GarageType":{np.nan: "None"},"GarageFinish":{np.nan: "None"},"GarageQual":{np.nan: "None"},"GarageCond":{np.nan: "None"}})
tt_copy['GarageType'].isnull().values.any()


False

In [35]:
# GarageYrBlt, GarageArea and GarageCars : Replacing missing data with 0 (Since No garage = no cars in such garage.)

tt_copy = tt_copy.replace({"GarageYrBlt":{np.nan: 0},"GarageArea":{np.nan: 0},"GarageCars":{np.nan: 0}})
tt_copy['GarageYrBlt'].isnull().values.any()
 

False

In [36]:
## BsmtFinSF1, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, BsmtFullBath and BsmtHalfBath : missing values are likely zero for having no basement

tt_copy = tt_copy.replace({"BsmtFinSF1":{np.nan: 0},"BsmtFinSF2":{np.nan: 0},"BsmtUnfSF":{np.nan: 0},"TotalBsmtSF":{np.nan: 0},"BsmtFullBath":{np.nan: 0},"BsmtHalfBath":{np.nan: 0}})
tt_copy['BsmtFinSF2'].isnull().values.any()

False

In [37]:
#BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1 and BsmtFinType2 : For all these categorical basement-related features, NaN means that there is no basement.

#tt_copy['BsmtQual'].isnull().values.any()
#tt_copy['BsmtQual'].unique()
#tt_copy['BsmtFinType2'].mode()

tt_copy = tt_copy.replace({"BsmtQual":{np.nan: "NB"},"BsmtCond":{np.nan: "NB"},"BsmtExposure":{np.nan: "NB"},"BsmtFinType1":{np.nan: "NB"},"BsmtFinType2":{np.nan: "NB"}})

#tt_copy['BsmtExposure'].unique()

#Here, 'NB' has been used to replace nan, and refers to No Basement condition

In [38]:
#tt_copy['BsmtCond'].isnull().values.any()
tt_copy['BsmtCond'].mode()

0    TA
dtype: object

In [39]:
#MasVnrArea and MasVnrType : NA most likely means no masonry veneer for these houses. We can fill 0 for the area and None for the type.
tt_copy['MasVnrArea'].isnull().values.any()

True

In [40]:
tt_copy['MasVnrArea'].isnull().sum()

23

In [41]:
tt_copy['MasVnrArea'].mode()

0    0.0
dtype: float64

In [42]:
tt_copy['MasVnrArea'] = tt_copy['MasVnrArea'].replace(np.nan,0)
tt_copy['MasVnrArea'].isnull().sum()

0

In [43]:
tt_copy['MasVnrType'].isnull().values.any()

True

In [44]:
tt_copy['MasVnrType'].isnull().sum()

24

In [45]:
tt_copy['MasVnrType'].mode()

0    None
dtype: object

In [46]:
tt_copy['MasVnrType'] = tt_copy['MasVnrType'].replace(np.nan,"None")
tt_copy['MasVnrType'].isnull().sum()

0

In [47]:
#MSZoning (The general zoning classification) : 'RL' is by far the most common value. So we can fill in missing values with 'RL'
tt_copy['MSZoning'].isnull().values.any()

True

In [48]:
tt_copy['MSZoning'].isnull().sum()

4

In [49]:
tt_copy['MSZoning'].mode()

0    RL
dtype: object

In [50]:
tt_copy['MSZoning'] = tt_copy['MSZoning'].replace(np.nan,"RL")

In [51]:
#Functional : data description says NA means typical
tt_copy['Functional'].isnull().values.any()

True

In [52]:
tt_copy['Functional'].isnull().sum()

2

In [53]:
tt_copy['Functional'].mode()

0    Typ
dtype: object

In [54]:
tt_copy['Functional'] = tt_copy['Functional'].replace(np.nan,"Typ")

In [55]:
#Electrical : It has one NA value. Since this feature has mostly 'SBrkr', we can set that for the missing value.
tt_copy['Electrical'].mode()

0    SBrkr
dtype: object

In [56]:
tt_copy['Electrical'].isnull().values.any()

True

In [57]:
tt_copy['Electrical'] = tt_copy['Electrical'].replace(np.nan,"SBrkr")

In [58]:
#KitchenQual: Only one NA value, and same as Electrical, we set 'TA' (which is the most frequent) for the missing value in KitchenQual.
tt_copy['KitchenQual'].mode()

0    TA
dtype: object

In [59]:
tt_copy['KitchenQual'].str.count("NA").sum()

0.0

In [60]:
tt_copy['KitchenQual'].isnull().values.any()

True

In [61]:
tt_copy['KitchenQual'] = tt_copy['KitchenQual'].replace(np.nan,"TA")

In [62]:
#Exterior1st and Exterior2nd : Again Both Exterior 1 & 2 have only one missing value. We will just substitute in the most common string
tt_copy['Exterior1st'].isnull().values.any()

True

In [63]:
tt_copy['Exterior1st'].mode()

0    VinylSd
dtype: object

In [64]:
tt_copy['Exterior1st'] = tt_copy['Exterior1st'].replace(np.nan,"VinylSd")

In [65]:
tt_copy['Exterior1st'].isnull().values.any()

False

In [66]:
tt_copy['Exterior2nd'].isnull().values.any()

True

In [67]:
tt_copy['Exterior2nd'].mode()

0    VinylSd
dtype: object

In [68]:
tt_copy['Exterior2nd'] = tt_copy['Exterior2nd'].replace(np.nan,"VinylSd")

In [69]:
tt_copy['Exterior2nd'].isnull().values.any()

False

In [70]:
#SaleType : Fill in again with most frequent which is "WD"
tt_copy['SaleType'].isnull().values.any()

True

In [71]:
tt_copy['SaleType'] = tt_copy['SaleType'].replace(np.nan,"WD")
tt_copy['SaleType'].isnull().values.any()

False

In [72]:
#MSSubClass : Na most likely means No building class. We can replace missing values with None
tt_copy['MSSubClass'].isnull().values.any()

False

### 1.17 Check again remaining missing values if any.And your observation?

In [73]:
tt_copy.isna().sum().sum()

1461

In [74]:
tt_copy.columns[tt_copy.isna().any()]

Index(['SalePrice', 'Utilities'], dtype='object')

In [75]:
tt_copy['SalePrice'].isnull().sum()

1459

In [76]:
tt_copy['Utilities'].isnull().sum()

2

In [77]:
tt_copy.dtypes['SalePrice'], tt_copy.dtypes['Utilities']

(dtype('float64'), dtype('O'))

In [78]:
tt_copy['Utilities'].unique()

array(['AllPub', 'NoSeWa', nan], dtype=object)

In [79]:
tt_copy['Utilities'] = tt_copy['Utilities'].transform(lambda x: x.fillna("NA"))
tt_copy['Utilities'].unique()

array(['AllPub', 'NoSeWa', 'NA'], dtype=object)

In [80]:
tt_copy['SalePrice'].describe()

count      1460.000000
mean     180921.195890
std       79442.502883
min       34900.000000
25%      129975.000000
50%      163000.000000
75%      214000.000000
max      755000.000000
Name: SalePrice, dtype: float64

In [81]:
Q1 = tt_copy['SalePrice'].quantile(0.25)
Q3 = tt_copy['SalePrice'].quantile(0.75)
IQR = Q3 - Q1
print("Outliers Number : ")
((tt_copy['SalePrice'] < (Q1 - 1.5 * IQR)) | (tt_copy['SalePrice'] > (Q3 + 1.5 * IQR))).sum()

Outliers Number : 


61

In [82]:
##Replace with mean instead of median since outliers are not many

In [83]:
tt_copy['SalePrice'] = tt_copy['SalePrice'].transform(lambda x: x.fillna(x.mean()))

#### observation:
    

In [84]:
tt_copy.isna().sum().sum()

0

In [85]:
##We could replace the missing values in the left over columns. 
##But, since imputation is the way to the goal and not the goal itself, 
##We maybe had to predict values for SalePrice column based on certain method, which is not asked in this assignment; 
##It only deals with missing values replacement, which is the goal for now

In [86]:
print("All missing values replaced")

All missing values replaced
